gateway/src/apicast/policy/standalone/standalone.lua

```diff
@@ -0,0 +1,364 @@
+    }, Route_mt)
+  end
+
+  function Route:match(context)
+    return self.conditions:all(context)
+  end
+end
+
+local Service = { }
+
+local function build_services(configuration)
+  return build_objects(Service.new, configuration.internal)
+end
+
+do
+  local Service_mt = { __index = Service }
+
+  local function build_policy_chain(policies)
+    local chain = tab_new(#policies, 0)
+
+    for i=1, #policies do
+      local policy, err = PolicyChain.load_policy(policies[i].policy, policies[i].version, policies[i].configuration)
+
+      if policy then
+        insert(chain, policy)
+      elseif err then
+        ngx.log(ngx.WARN, 'failed to load policy: ', policies[i].policy, ' version: ', policies[i].version)
+        ngx.log(ngx.DEBUG, err)
+      end
+    end
+
+    return PolicyChain.new(chain)
+  end
+
+  function Service.new(config)
+    return setmetatable({
+      name = config.name,
+      upstream = Upstream.new(config.upstream),
+      policy_chain = build_policy_chain(config.policy_chain),
+    }, Service_mt)
+  end
+end
+
+local External = { }
+
+local function build_upstreams(configuration)
+  return build_objects(External.new, configuration.external)
+end
+
+do
+  local External_mt = { __index = External }
+
+  function External.new(config)
+    return setmetatable({
+      name = config.name,
+      server = Upstream.new(config.server),
+      load_balancer = config.load_balancer,
+      retries = config.retries,
+    }, External_mt)
+  end
+end
+
+local default = {
+  services = build_objects(Service.new, {
+    { name = 'not_found',
+      policy_chain = {
+        { policy = 'apicast.policy.echo', configuration = { status = ngx.HTTP_NOT_FOUND } },
+      },
+    },
+  }),
+}
+
+function _M:load_configuration()
+  local url = self.url
+
+  if not url then
+    return nil, 'not initialized'
+  end
+
+  local configuration, err = load_configuration(url)
+
+  if configuration then
+    self.routes = build_routes(configuration)
+    self.services = setmetatable(build_services(configuration), { __index = default.services })
+    self.upstreams = build_upstreams(configuration)
+
+    ngx.log(ngx.NOTICE, 'loaded standalone configuration from: ', url)
+    return configuration
+  else
+    ngx.log(ngx.WARN, 'failed to load ', url, ' err: ', err)
+
+    self.routes = {}
+    self.services = setmetatable({}, { __index = default.services })
+    self.upstreams = {}
+
+    return nil, err
+  end
+end
+
+local function run_phase(phase, services, ...)
+  if not services then return end
+
+  for _, service in ipairs(services) do
+    ngx.log(ngx.DEBUG, 'running phase ', phase, ' on service ', service.name)
+    service.policy_chain[phase](service.policy_chain, ...)
+  end
+end
+
+function _M:init(...)
+  if self then -- initializing policy instance
+    local config, err = self:load_configuration(self)
+
+    if config then
+      -- TODO: we need to run this because some policies are "internal" and not being
```

Comment:
I think we had a similar problem in `Executor` and figure out a way to run every phase only once.

Reply:
It was a bit different problem I think.

Executor does the following:

* call `:init` and initialize all policies in the current chain
* use `PolicyLoader:get_all()` to load "all" policies
* call `:init` on the remaining policies

Now  `PolicyLoader:get_all()` does not return all the policies. Some don't have `apicast-policy.json` file, so they are not discovered.

So the problem is that the Executor is going to call the `:init` phase on all policies that it can find but does not expose the information if the phase was already executed. So standalone can't really know if it was already executed or not.

There are two options I see:

* expose information about what phase was executed on what policy in the Executor
* standalone policy should not trigger `:init` on policies that can be found by `PolicyLoader` and assume it was done by the Executor module


Wdyt?

Reply:
I prefer the first one because I think it's more explicit, but I think that both options should be OK.

Reply:
So after some digging, I realised the problem is a bit deeper.

Executor keeps track of what it executed, but it can't keep track of what nested policy chains executed. Let's say Executor has 3 policies and the 2nd is itself a policy chain with 3 more policies. The Executor would be aware only of the first level. 

We would have to figure out how to collect all policies from the executor without calling them.


Maybe I'd rather postpone this until we get more information and see more places where it breaks.

Reply:
Ah, I didn't think about the nested chains scenario.
OK to postpone it :+1: No need to solve this for a first version.